File: sdks/node/src/data/storage/ipfs.ts
import {
  type DataStorage,
  type FileMetadata,
  type GetShareLinkOptions,
  StorageError,
  StorageType,
  type UploadOptions,
} from "./interfaces";

export const IPFS_LINK: string = "https://ipfs.io";
export const IPFS_DWEB_LINK: string = "https://dweb.link";
export const IPFS_W3S_LINK: string = "https://w3s.link";
export const IPFS_TRUSTLESS_GATEWAY_LINK: string =
  "https://trustless-gateway.link";
export const IPFS_4EVERLAND_LINK: string = "https://4everland.io";
export const IPFS_PINATA_CLOUD_LINK: string = "https://gateway.pinata.cloud";
export const IPFS_NFT_STORAGE_LINK: string = "https://nftstorage.link";

export const IPFS_GATEWAY_ENV: string = "IPFS_GATEWAY";
export const IPFS_API_KEY_ENV: string = "IPFS_API_KEY";
export const IPFS_API_SECRET_ENV: string = "IPFS_API_SECRET_KEY";
export const IPFS_JWT_ENV: string = "IPFS_JWT";

export class PinataFileDetails {
  acceptDuplicates: boolean;
  isDuplicate: boolean | null;
  id: string;
  userId: string | null;
  name: string;
  cid: string;
  size: number;
  numberOfFiles: number;
  mimeType: string;
  groupId: string | null;
  createdAt: string;
  updatedAt: string;
  network: string;
  streamable: boolean;
  vectorized: boolean | null;

  constructor() {
    this.acceptDuplicates = false;
    this.isDuplicate = false;
    this.id = "";
    this.userId = "";
    this.name = "";
    this.cid = "";
    this.size = 0;
    this.numberOfFiles = 0;
    this.mimeType = "";
    this.groupId = null;
    this.createdAt = "";
    this.updatedAt = "";
    this.network = "";
    this.streamable = false;
    this.vectorized = false;
  }
}

export class PinataUploadResponse {
  data: PinataFileDetails;

  constructor() {
    this.data = new PinataFileDetails();
  }
}

export class PinataIPFS implements DataStorage {
  static new(): PinataIPFS {
    return new PinataIPFS();
  }

  async upload(opts: UploadOptions): Promise<FileMetadata> {
    const { name, data, token } = opts;
    const url = "https://uploads.pinata.cloud/v3/files";

    const formData = new FormData();
    const filePart = new Blob([data.toString()], { type: "text/plain" });
    formData.append("file", filePart, name);
    formData.append("network", "public");

    try {
      const response = await fetch(url, {
        method: "POST",
        body: formData,
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new StorageError(`Pinata IPFS API error: ${errorText}`);
      }

      const resp = (await response.json()) as PinataUploadResponse;
      return PinataFileDetailsToMetadata(resp.data);
    } catch (error) {
      throw new StorageError(`Upload error: ${error}`);
    }
  }

  async getShareLink(opts: GetShareLinkOptions): Promise<string> {
    return `https://gateway.pinata.cloud/ipfs/${opts.id}?download=true`;
  }

  storageType(): StorageType {
    return StorageType.IPFS;
  }
}

function PinataFileDetailsToMetadata(
  pinataFileDetails: PinataFileDetails
): FileMetadata {
  return {
    id: pinataFileDetails.cid,
    name: pinataFileDetails.name,
    size: pinataFileDetails.size,
    modifiedTime: pinataFileDetails.updatedAt || null,
  };
}

File: crates/data/src/storage/ipfs.rs
use crate::storage::{DataStorage, FileMetadata, GetShareLinkOptions, StorageType, UploadOptions};
use anyhow::{Context, Result};
use reqwest::{Client, multipart};
use serde::{Deserialize, Serialize};

// IPFS Gateway Links

pub const IPFS_LINK: &str = "https://ipfs.io";
pub const IPFS_DWEB_LINK: &str = "https://dweb.link";
pub const IPFS_W3S_LINK: &str = "https://w3s.link";
pub const IPFS_TRUSTLESS_GATEWAY_LINK: &str = "https://trustless-gateway.link";
pub const IPFS_4EVERLAND_LINK: &str = "https://4everland.io";
pub const IPFS_PINATA_CLOUD_LINK: &str = "https://gateway.pinata.cloud";
pub const IPFS_NFT_STORAGE_LINK: &str = "https://nftstorage.link";

// IPFS Secret Env Vars

pub const IPFS_GATEWAY_ENV: &str = "IPFS_GATEWAY";
pub const IPFS_API_KEY_ENV: &str = "IPFS_API_KEY";
pub const IPFS_API_SECRET_ENV: &str = "IPFS_API_SECRET_KEY";
pub const IPFS_JWT_ENV: &str = "IPFS_JWT";

#[derive(Debug, Clone, Default)]
pub struct PinataIPFS {
    client: Client,
}

impl PinataIPFS {
    pub fn new<S: AsRef<str>>() -> Self {
        Self {
            client: Client::new(),
        }
    }

    pub async fn get_share_link<S: AsRef<str>>(&self, _token: S, cid: S) -> Result<String> {
        Ok(format!(
            "https://gateway.pinata.cloud/ipfs/{}?download=true",
            cid.as_ref()
        ))
    }
}

#[async_trait::async_trait]
impl DataStorage for PinataIPFS {
    async fn upload(&self, opts: UploadOptions) -> Result<FileMetadata> {
        let UploadOptions { name, data, token } = opts;
        let url = "https://uploads.pinata.cloud/v3/files";
        let file_part = multipart::Part::text(String::from_utf8(data)?)
            .file_name(name)
            .mime_str("text/plain")?;
        let form = multipart::Form::new()
            .part("file", file_part)
            .text("network", "public");
        let response = self
            .client
            .post(url)
            .multipart(form)
            .bearer_auth(token.as_str())
            .send()
            .await
            .context("Failed to send upload request")?;
        if !response.status().is_success() {
            let error_text = response.text().await?;
            anyhow::bail!("Pinata IPFS API error: {}", error_text);
        }
        let resp = response.json::<PinataUploadResponse>().await?;
        Ok(resp.data.into())
    }

    async fn get_share_link(&self, opts: GetShareLinkOptions) -> Result<String> {
        let GetShareLinkOptions { token, id } = opts;
        Ok(self.get_share_link(token, id).await?)
    }

    #[inline]
    fn storage_type(&self) -> StorageType {
        StorageType::IPFS
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct PinataUploadResponse {
    pub data: PinataFileDetails,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct PinataFileDetails {
    pub accept_duplicates: Option<bool>,
    pub is_duplicate: Option<bool>,
    pub id: String,
    pub user_id: Option<String>,
    pub name: String,
    pub cid: String,
    pub size: usize,
    pub number_of_files: usize,
    pub mime_type: String,
    pub group_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub network: String,
    pub streamable: bool,
    pub vectorized: Option<bool>,
}

impl From<PinataFileDetails> for FileMetadata {
    fn from(value: PinataFileDetails) -> Self {
        FileMetadata {
            id: value.cid,
            name: value.name,
            size: value.size,
            modified_time: Some(value.updated_at),
        }
    }
}
